polkadot/xcm/xcm-fee-payment-runtime-api/tests/fee_estimation.rs

```diff
@@ -0,0 +1,262 @@
+// Copyright (C) Parity Technologies (UK) Ltd.
+// This file is part of Polkadot.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Tests for using both the XCM fee payment API and the transfers API.
+
+use sp_api::ProvideRuntimeApi;
+use sp_runtime::testing::H256;
+use xcm::prelude::*;
+use xcm_fee_payment_runtime_api::{XcmDryRunApi, XcmPaymentApi};
+
+mod mock;
+use mock::{
+	extra, new_test_ext_with_balances, new_test_ext_with_balances_and_assets, DeliveryFees,
+	ExistentialDeposit, HereLocation, RuntimeCall, TestClient, TestXt,
+};
+
+// Scenario: User `1` in the local chain wants to transfer assets to account `[0u8; 32]` on
+// "AssetHub". He wants to make sure he has enough for fees, so before he calls the `transfer_asset`
+// extrinsic to do the transfer, he decides to use the `XcmDryRunApi` and `XcmPaymentApi` runtime
+// APIs to estimate fees. This uses a teleport because we're dealing with the native token of the
+// chain, which is registered on "AssetHub".
+#[test]
+fn fee_estimation_for_teleport() {
```

Comment:
Would be nice to transfer both `NativeToken` and `RelayToken` for a more comprehensive scenario.